NEW: Save and load rebinds as JSON. (#1186)
Unity-Technologies/InputSystem: CONTRIBUTIONS.md

# Contributions

## If you are interested in contributing, here are some ground rules:
* Ideally, [talk to us](https://forum.unity.com/forums/input-system.103/) before doing the work -- we appreciate contributions, but we might already be working on the same thing, or we might have different opinions on how it should be implemented.

> ## All contributions are subject to the [Unity Contribution Agreement(UCA)](https://unity3d.com/legal/licenses/Unity_Contribution_Agreement)

Reviewer: Does this come from Legal? Its worth getting them to check and make sure the wording is correct.

Author: This one is copied from the various other CONTRIBUTIONS.md we have. Probably a bit cheap on my end but with some of that stuff already on trunk and with the doc itself up on unity3d.com/legal, I just told myself "this is already properly vetted"  :)

By making a pull request, you are confirming agreement to the terms and conditions of the UCA, including that your Contributions are your original creation and that you have complete right and authority to make your Contributions.

> ## Once you have a change ready following these ground rules. Simply make a pull request in Github

Reviewer 2: ```suggestion
## Once you have a change ready that complies to the rules above, open a pull request in GitHub.

```

Author: Pushed

> Some notes:

Reviewer 2: ```suggestion
Notes:

```

Author: Pushed

> * We can rarely take a pull request directly and as is. The reason for this is that our CI will not be run on PRs coming from forks. So, in most cases we will have to incorporate your changes into a branch inside the repo and PR it into the `develop` branch via a separate PR. When we do so, we will clearly state that in your PR.

Reviewer 2: ```suggestion
* We can rarely take a pull request directly and as-is because our CI doesn't run on PRs coming from forks. So, in most cases we will have to incorporate your changes into a branch inside the repository and merge it into the `develop` branch via a separate PR. When we do so, we will clearly state that in your PR.

```

Reviewer 2: Do we need to tell them why we can't take PRs directly? The bit about our CI not running on PRs coming from forks seems superfluous to me (but that might just be my minimalist brain at work).

Author: I think that part does matter. Making a pull request, I'd expect someone to either take it or reject it. Whereas in our current workflow, even if we do take it, it will end up getting closed and not merged and your contribution won't really show as such in the history.

Author: Pushed

* All changes are subject to the following requirements:
  1) Tests need to be green.

>   2) Behavioral changes need to be accompanied by test changes/additions.

Reviewer 2: ```suggestion
  2) Behavioral changes need to be accompanied by test changes and additions.

```

Reviewer 2: Do we mean that tests need to cover any behavioral changes? If so, we could probably rephrase this to make it clearer.

>   3) Where relevant, documentation (xmldoc and/or manual) needs to be updated.

Reviewer 2: ```suggestion
  3) Where relevant, documentation (xmldoc and/or Manual) needs to be updated.

```

Author: Pushed

  4) Fixes, changes, and additions need to be documented in [CHANGELOG](./CHANGELOG.md).
  5) Breaking API changes can only be introduced in new major versions.
* The more a PR deviates from these requirements, the more work we have to do on our side to incorporate your changes. Give your PR a leg up by making it easy on us :)

> * We are hoping to have better/easier flows for incorporating user PRs in the future.

Reviewer 2: ```suggestion
* We are hoping to have an easier workflow for incorporating user PRs in the future.
```

Reviewer 2: Just a style edit to avoid the slash.

Author: Pushed
